File: nodejs-soa/order-service/src/infrastructure/messaging/events.ts
// =============================================================================
// Event Definitions - Order Service
// =============================================================================
// Defines the events that the Order Service publishes and consumes.
//
// Published Events:
// - order.created: When a new order is placed (triggers Saga)
// - order.confirmed: When stock is reserved and order is confirmed
// - order.cancelled: When order is cancelled (triggers compensation)
//
// Consumed Events (from Inventory Service):
// - stock.reserved: Stock was successfully reserved
// - stock.insufficient: Not enough stock available
// =============================================================================

import { Order } from '../../domain/model/Order';

// --- Published Events ---

export interface OrderCreatedEvent {
  eventType: 'ORDER_CREATED';
  timestamp: string;
  data: {
    orderId: string;
    userId: string;
    items: Array<{
      productId: string;
      productName: string;
      quantity: number;
      unitPrice: number;
    }>;
    totalAmount: number;
  };
}

export interface OrderConfirmedEvent {
  eventType: 'ORDER_CONFIRMED';
  timestamp: string;
  data: {
    orderId: string;
    userId: string;
    totalAmount: number;
  };
}

export interface OrderCancelledEvent {
  eventType: 'ORDER_CANCELLED';
  timestamp: string;
  data: {
    orderId: string;
    userId: string;
    items: Array<{
      productId: string;
      quantity: number;
    }>;
  };
}

// --- Consumed Events ---

export interface StockReservedEvent {
  eventType: 'STOCK_RESERVED';
  timestamp: string;
  data: {
    orderId: string;
    items: Array<{
      productId: string;
      quantityReserved: number;
    }>;
  };
}

export interface StockInsufficientEvent {
  eventType: 'STOCK_INSUFFICIENT';
  timestamp: string;
  data: {
    orderId: string;
    productId: string;
    requestedQuantity: number;
    availableQuantity: number;
  };
}

// --- Factory Functions ---

export function createOrderCreatedEvent(order: Order): OrderCreatedEvent {
  return {
    eventType: 'ORDER_CREATED',
    timestamp: new Date().toISOString(),
    data: {
      orderId: order.id,
      userId: order.userId,
      items: order.items.map((item) => ({
        productId: item.productId,
        productName: item.productName,
        quantity: item.quantity,
        unitPrice: item.unitPrice,
      })),
      totalAmount: order.totalAmount,
    },
  };
}

export function createOrderConfirmedEvent(order: Order): OrderConfirmedEvent {
  return {
    eventType: 'ORDER_CONFIRMED',
    timestamp: new Date().toISOString(),
    data: {
      orderId: order.id,
      userId: order.userId,
      totalAmount: order.totalAmount,
    },
  };
}

export function createOrderCancelledEvent(order: Order): OrderCancelledEvent {
  return {
    eventType: 'ORDER_CANCELLED',
    timestamp: new Date().toISOString(),
    data: {
      orderId: order.id,
      userId: order.userId,
      items: order.items.map((item) => ({
        productId: item.productId,
        quantity: item.quantity,
      })),
    },
  };
}

File: python-soa/order-service/app/infrastructure/messaging/events.py
"""
Order Events
==============
Factory methods for creating order-related domain events.

Events Published:
- order.created: Initiates the saga (consumed by Inventory Service)
- order.confirmed: Saga completed successfully
- order.cancelled: Compensation event (triggers stock release)
"""

from datetime import datetime

from app.domain.model import Order


class OrderEvents:
    """Factory class for creating order-related domain events."""

    @staticmethod
    def order_created(order: Order) -> dict:
        """
        Create an 'order.created' event payload.

        This event initiates the order saga. The Inventory Service
        consumes it and attempts to reserve stock for all items.

        Contains all item details needed for stock reservation.
        """
        return {
            "event_type": "order.created",
            "timestamp": datetime.utcnow().isoformat(),
            "data": {
                "order_id": str(order.id),
                "user_id": str(order.user_id),
                "items": [
                    {
                        "product_id": str(item.product_id),
                        "product_name": item.product_name,
                        "quantity": item.quantity,
                        "unit_price": float(item.unit_price),
                    }
                    for item in order.items
                ],
                "total_amount": float(order.total_amount),
            },
        }

    @staticmethod
    def order_confirmed(order: Order) -> dict:
        """
        Create an 'order.confirmed' event payload.

        Published when the saga completes successfully.
        Can be consumed by notification services, analytics, etc.
        """
        return {
            "event_type": "order.confirmed",
            "timestamp": datetime.utcnow().isoformat(),
            "data": {
                "order_id": str(order.id),
                "user_id": str(order.user_id),
                "total_amount": float(order.total_amount),
                "status": order.status.value,
            },
        }

    @staticmethod
    def order_cancelled(order: Order, reason: str = "") -> dict:
        """
        Create an 'order.cancelled' event payload.

        Published when a confirmed order is cancelled.
        The Inventory Service consumes this to release reserved stock.
        """
        return {
            "event_type": "order.cancelled",
            "timestamp": datetime.utcnow().isoformat(),
            "data": {
                "order_id": str(order.id),
                "user_id": str(order.user_id),
                "reason": reason,
                "items": [
                    {
                        "product_id": str(item.product_id),
                        "quantity": item.quantity,
                    }
                    for item in order.items
                ],
            },
        }
